lib/SILOptimizer/Mandatory/MoveOnlyDiagnostics.cpp

```diff
@@ -190,6 +191,125 @@ void DiagnosticEmitter::emitCheckedMissedCopyError(SILInstruction *copyInst) {
            diag::sil_movechecking_bug_missed_copy);
 }
 
+void DiagnosticEmitter::emitMissingConsumeInDiscardingContext(
+    SILInstruction *leftoverDestroy,
+    SILInstruction *discard) {
+  assert(isa<DropDeinitInst>(discard));
+
+  // A good location is one that has some connection with the original source
+  // and corresponds to an exit of the function.
+  auto hasGoodLocation = [](SILInstruction *si) -> bool {
+    if (!si)
+      return false;
+
+    SILLocation loc = si->getLoc();
+    if (loc.isNull())
+      return false;
+
+    switch (loc.getKind()) {
+    case SILLocation::ReturnKind:
+    case SILLocation::ImplicitReturnKind:
+      return true;
+
+    case SILLocation::RegularKind: {
+      Stmt *stmt = loc.getAsASTNode<Stmt>();
+      if (!stmt)
+        return true; // For non-statements, assume it is exiting the func.
+
+      // Prefer statements that can possibly lead to an exit of the function.
+      // This is determined by whether the statement causes an exit of a
+      // lexical scope; so a 'break' counts but not a 'continue'.
+      switch (stmt->getKind()) {
+      case StmtKind::Throw:
+      case StmtKind::Return:
+      case StmtKind::Yield:
+      case StmtKind::Break:
+      case StmtKind::Fail:
+      case StmtKind::PoundAssert:
+        return true;
+
+      case StmtKind::Continue:
+      case StmtKind::Brace:
+      case StmtKind::Defer:
+      case StmtKind::If:
+      case StmtKind::Guard:
+      case StmtKind::While:
+      case StmtKind::Do:
+      case StmtKind::DoCatch:
+      case StmtKind::RepeatWhile:
+      case StmtKind::ForEach:
+      case StmtKind::Switch:
+      case StmtKind::Case:
+      case StmtKind::Fallthrough:
+      case StmtKind::Discard:
+        return false;
+      };
+    }
+
+    case SILLocation::InlinedKind:
+    case SILLocation::MandatoryInlinedKind:
+    case SILLocation::CleanupKind:
+    case SILLocation::ArtificialUnreachableKind:
+      return false;
+    };
+  };
+
+  // An instruction corresponding to the logical place where the value is
+  // destroyed. Ideally an exit point of the function reachable from here or
+  // some relevant statement.
+  SILInstruction *destroyPoint = leftoverDestroy;
+  if (!hasGoodLocation(destroyPoint)) {
+    // Search for a nearby function exit reachable from this destroy. We do this
+    // because the move checker may have injected or hoisted an existing
+    // destroy from leaf blocks to some earlier point. For example, if 'd'
+    // represents a destroy of self, then we may have this CFG:
+    //
+    //        before:      after:
+    //           .          d
+    //          / \        / \
+    //         d   d      .   .
+    //
+    BasicBlockSet visited(destroyPoint->getFunction());
+    std::deque<SILBasicBlock *> bfsWorklist = {destroyPoint->getParent()};
```

Comment:
We generally do not use std::deque in the code base. Does this really need to be a BFS? Couldn't it be a DFS and you use a BasicBlockWorklist? If you really need a queue, my suggestion is to use two std::vectors and swap them.

Reply:
Reading this, I am pretty sure that this doesn't need a DFS. Can you use a BasicBlockWorklist here? It should be more efficient and we use it throughout the optimizer in this situation.

Reply:
I did see uses of std::deque for BFS in the code base. As the comment mentions, I'm looking for the nearest exit of the function so a DFS will not work. I don't see a reason to overcomplicate this queue in an effort to optimize it, when it cannot be a performance issue: this code is for emitting an error diagnostic.

Reply:
Does it really matter if we emit the closest one in terms of a CFG walk? The notion of closeness in terms of the CFG doesn't necessarily correspond to closeness in the source code (which is what I believe you are trying to approximate, no?). For example consider a situation where one has a split CFG like the followings:

```
   bb0
  /        \
v.         v
bb1    bb2
            |
           v
          bb3
```

where bb1 is a huge basic block containing 5 pages of code and bb2/bb3 are two small blocks due to an early exit. In this case a BFS would give an apparent location far away from where-ever we are emitting errors from. I have definitely seen cases like this in real world code. If one accepts that CFG closeness correspond to closeness in source code, whether or not one uses a DFS or a BFS becomes immaterial. 

In terms of using deque, this is about using standardized patterns throughout the optimizer/code base. In general, we have a set of standardized ways of doing things in the code base. One of them is when iterating over blocks to use BasicBlockWorklist if possible. It isn't just about performance (even though that is nice)... it is also about standardizing our code into standard patterns/utilities.

Reply:
That's a good point. Neither approach is perfect and some other strategy is probably better. I settled on BFS because in practice it gives better locations, for example, from one of the regression tests:

```
struct Basics: ~Copyable {
  borrowing func thrower() throws {}
  consuming func test12(_ c: Color) throws { // with DFS the error ends up here
    guard case .red = c else {
      discard self
      return
    }
    try thrower() // with BFS the error is nicely located here
    print("hi")
    _ = consume self
  }
```

This is an important situation, since it's not always obvious to people that a `try` is a possible function exit.

Reply:
I am sure I can come up with a similar test case where one could get the other behavior. That being said, it seems like you feel very strongly about this. I am fine with this.